imap_processing/cli.py

```diff
@@ -241,67 +243,159 @@ def download_dependencies(self):
             file_list.append(imap_data_access.download(return_query[0]["file_path"]))
         return file_list
 
-    @abstractmethod
+    def upload_products(self, products: list[str]):
+        """
+        Upload data products to the IMAP SDC.
+
+        Attributes
+        ----------
+        products: list[str]
+        A list of file paths to upload to the SDC.
+        """
+        if self.upload_to_sdc:
+            if len(products) == 0:
+                logger.info("No files to upload.")
+            for filename in products:
+                logger.info(f"Uploading file: {filename}")
+                imap_data_access.upload(filename)
+
+    @final
     def process(self):
-        """Perform instrument specific processing."""
+        """
+        Run the processing workflow and cannot be overridden by subclasses.
+
+        Each IMAP processing step consists of three steps:
+        1. Pre-processing actions such as downloading dependencies for processing.
+        2. Do the data processing. The result of this step will usually be a list
+        of new products (files).
+        3. Post-processing actions such as uploading files to the IMAP SDC.
+        """
+        dependencies = self.pre_processing()
+        products = self.do_processing(dependencies)
+        self.post_processing(products)
+
+    def pre_processing(self):
+        """
+        Complete pre-processing.
+
+        For this baseclass, pre-processing consists of downloading
+        dependencies for processing. Child classes can override this
+        method to customize the pre-processing actions.
+
+        Returns
+        -------
+        List of dependencies downloaded from the IMAP SDC.
+        """
```

Comment:
Also super nit-picky but I think we should strive for consistency in using up the 80-character limit in our docstrings

```suggestion
        """
        Complete pre-processing.

        For this baseclass, pre-processing consists of downloading dependencies
        for processing. Child classes can override this method to customize the
        pre-processing actions.

        Returns
        -------
        List of dependencies downloaded from the IMAP SDC.
        """
```

Reply:
Not super nit-picky. Thanks for suggesting this.